src/crypto/hmac.rs

```diff
@@ -1,106 +1,161 @@
+            .map_err(|_e| crate::errors::ErrorKind::InvalidKeyFormat)?;
+
+        Ok(Self(inner))
     }
-    fn verify(&mut self, signature: &[u8], message: &[u8]) -> bool {
-        self.reset();
-        self.update(message);
-        self.clone().verify_slice(signature).is_ok()
+}
+
+impl Signer<Vec<u8>> for Hs384 {
+    fn try_sign(&self, msg: &[u8]) -> std::result::Result<Vec<u8>, signature::Error> {
+        let mut signer = self.0.clone();
+        signer.reset();
+        signer.update(msg);
+
+        Ok(signer.finalize().into_bytes().to_vec())
     }
 }
 
-impl HmacAlgorithm for HmacSha384 {
-    fn sign(&mut self, message: &[u8]) -> Vec<u8> {
-        self.reset();
-        self.update(message);
-        self.clone().finalize().into_bytes().to_vec()
+impl JwtSigner for Hs384 {
+    fn algorithm(&self) -> Algorithm {
+        Algorithm::HS384
     }
-    fn verify(&mut self, signature: &[u8], message: &[u8]) -> bool {
-        self.reset();
-        self.update(message);
-        self.clone().verify_slice(signature).is_ok()
+}
+
+impl Verifier<Vec<u8>> for Hs384 {
+    fn verify(&self, msg: &[u8], signature: &Vec<u8>) -> std::result::Result<(), signature::Error> {
+        let mut verifier = self.0.clone();
+        verifier.reset();
+        verifier.update(msg);
+
+        verifier.verify_slice(signature).map_err(|e| signature::Error::from_source(e))
+    }
+}
+
+impl JwtVerifier for Hs384 {
+    fn algorithm(&self) -> Algorithm {
+        Algorithm::HS384
     }
 }
 
-impl HmacAlgorithm for HmacSha512 {
-    fn sign(&mut self, message: &[u8]) -> Vec<u8> {
-        self.reset();
-        self.update(message);
-        self.clone().finalize().into_bytes().to_vec()
+pub struct Hs512(HmacSha512);
+
+impl Hs512 {
+    pub(crate) fn new(secret: HmacSecret) -> Result<Self> {
+        let inner = HmacSha512::new_from_slice(&secret.0)
+            .map_err(|_e| crate::errors::ErrorKind::InvalidKeyFormat)?;
+
+        Ok(Self(inner))
     }
+}
+
+impl Signer<Vec<u8>> for Hs512 {
+    fn try_sign(&self, msg: &[u8]) -> std::result::Result<Vec<u8>, signature::Error> {
+        let mut signer = self.0.clone();
+        signer.reset();
+        signer.update(msg);
+
+        Ok(signer.finalize().into_bytes().to_vec())
+    }
+}
+
+impl JwtSigner for Hs512 {
+    fn algorithm(&self) -> Algorithm {
+        Algorithm::HS512
+    }
+}
+
+impl Verifier<Vec<u8>> for Hs512 {
+    fn verify(&self, msg: &[u8], signature: &Vec<u8>) -> std::result::Result<(), signature::Error> {
+        let mut verifier = self.0.clone();
+        verifier.reset();
+        verifier.update(msg);
+
+        verifier.verify_slice(signature).map_err(|e| signature::Error::from_source(e))
+    }
+}
 
-    fn verify(&mut self, signature: &[u8], message: &[u8]) -> bool {
-        self.reset();
-        self.update(message);
-        self.clone().verify_slice(signature).is_ok()
+impl JwtVerifier for Hs512 {
+    fn algorithm(&self) -> Algorithm {
+        Algorithm::HS512
```

Comment:
Looks like the kind of code that could be simplified with macros

Reply:
For repeated things that that, sure it could :+1: 

I feel that we are only repeating it a couple times so I don't believe the complexity of a macro is worth it. I also find it easier to read normal Rust code. But maybe that is just me :shrug: